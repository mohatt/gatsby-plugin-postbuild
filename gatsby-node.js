export * from './dist/gatsby-node.cjs'
