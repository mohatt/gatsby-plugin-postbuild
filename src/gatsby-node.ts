import Postbuild from './postbuild'
import { createPluginExport, PLUGIN, reporter } from './common'
import MinifyTask from './tasks/minify'
import PurgeCssTask from './tasks/purgecss'
import HttpHeadersTask from './tasks/http-headers'

// The main plugin object that handles core functionality
const postbuild = new Postbuild()

// Register core tasks
postbuild.init([MinifyTask, PurgeCssTask, HttpHeadersTask])

// Validates user-defined options against schema (runs before onPluginInit)
export const pluginOptionsSchema = createPluginExport('pluginOptionsSchema', ({ Joi }) => {
  return postbuild.getOptionsSchemas(Joi)
})

// Initializes plugin state
export const onPluginInit = createPluginExport('onPluginInit', async (args, pluginOptions) => {
  args.emitter.on('HTML_GENERATED', (action) => {
    postbuild.setBuildPaths(action.payload)
  })
  await postbuild.bootstrap(args, pluginOptions)
})

// Sets webpack config for the current stage
export const onCreateWebpackConfig = createPluginExport(
  'onCreateWebpackConfig',
  ({ actions, stage }) => {
    actions.setWebpackConfig(postbuild.getWebpackConfig(stage as string))
  },
)

/**
 * Runs Postbuild on the generated static files
 */
export const onPostBuild = createPluginExport('onPostBuild', async (args) => {
  const activity = args.reporter.activityTimer(PLUGIN, {
    // @ts-expect-error
    parentSpan: args.tracing.parentSpan,
  })
  activity.start()
  try {
    // @todo need to check for incremental build before running postbuild
    await postbuild.run(args, activity.setStatus)
  } catch (e) {
    activity.panic(reporter.createError('onPostBuild', e))
  }
  activity.end()
})
