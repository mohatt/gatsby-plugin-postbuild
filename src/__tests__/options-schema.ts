import { testPluginOptionsSchema } from 'gatsby-plugin-utils'
import { pluginOptionsSchema } from '@postbuild/gatsby-node'
import optionsFixtures from './__fixtures__/options'

describe('options schema', () => {
  it.each(optionsFixtures)('$title', async ({ options }) => {
    const { isValid, errors } = await testPluginOptionsSchema(pluginOptionsSchema, options as any)
    expect(isValid).toMatchSnapshot()
    expect(errors).toMatchSnapshot()
  })
})
