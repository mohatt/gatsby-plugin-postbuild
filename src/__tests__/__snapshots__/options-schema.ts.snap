// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`options schema > 'should invalidate empty options' 1`] = `false`;

exports[`options schema > 'should invalidate empty options' 2`] = `
[
  ""ignore[0]" is not allowed to be empty",
  ""processing.concurrency" must be greater than or equal to 1",
  ""processing.strategy" must be one of [sequential, parallel]",
  ""processing.strategy" is not allowed to be empty",
  ""extensions.txt.concurrency" must be greater than or equal to 1",
  ""extensions.txt.strategy" must be one of [sequential, parallel]",
  ""extensions.txt.strategy" is not allowed to be empty",
]
`;

exports[`options schema > 'should invalidate invalid options' 1`] = `false`;

exports[`options schema > 'should invalidate invalid options' 2`] = `
[
  ""events.on.load" is not allowed",
  ""events.html.content" is not allowed",
  ""processing.concurrency" must be greater than or equal to 1",
  ""processing.strategy" must be one of [sequential, parallel]",
  ""extensions.txt.concurrency" must be greater than or equal to 1",
  ""extensions.txt.strategy" must be one of [sequential, parallel]",
]
`;

exports[`options schema > 'should invalidate invalid tasks options' 1`] = `false`;

exports[`options schema > 'should invalidate invalid tasks options' 2`] = `
[
  ""http-headers.provider" must be one of [netlify, vercel, firebase]",
  ""http-headers.headers" must be of type object",
  ""http-headers.caching" must be a boolean",
  ""minify.style" must be one of [boolean, object, array, function]",
  ""minify.enabled" must be a boolean",
  ""minify.ignore" must be an array",
]
`;

exports[`options schema > 'should invalidate options with incorrect data types' 1`] = `false`;

exports[`options schema > 'should invalidate options with incorrect data types' 2`] = `
[
  ""enabled" must be a boolean",
  ""reporting" must be one of [object, boolean]",
  ""ignore" must be an array",
  ""events.on.bootstrap" must be of type function",
  ""events.html.parse" must be of type function",
  ""events.foo" must be of type object",
  ""events.txt.content" must be of type function",
  ""processing.concurrency" must be a number",
  ""processing.strategy" must be one of [sequential, parallel]",
  ""processing.strategy" must be a string",
  ""extensions.foo" must be of type object",
  ""extensions.txt.concurrency" must be a number",
  ""extensions.txt.strategy" must be one of [sequential, parallel]",
  ""extensions.txt.strategy" must be a string",
]
`;

exports[`options schema > 'should validate correct options (with alternatives)' 1`] = `true`;

exports[`options schema > 'should validate correct options (with alternatives)' 2`] = `[]`;

exports[`options schema > 'should validate correct options' 1`] = `true`;

exports[`options schema > 'should validate correct options' 2`] = `[]`;

exports[`options schema > 'should validate correct tasks options' 1`] = `true`;

exports[`options schema > 'should validate correct tasks options' 2`] = `[]`;

exports[`options schema > 'should validate default options' 1`] = `true`;

exports[`options schema > 'should validate default options' 2`] = `[]`;
